02-peer-protocol.md

```diff
@@ -1515,6 +1522,51 @@ fall-behind detection.  An implementation can offer both, however, and
 fall back to the `option_data_loss_protect` behavior if
 `option_static_remotekey` is not negotiated.
 
+## Channel Backup Storage
+
+Nodes that advertise the `provide_peer_backup_storage` feature offer storing
+arbitrary data for their peers. [BOLT #1](01-messaging.md#node-backup-storage)
+describes how this allows storing per-node data, and the following section
+describes how this allows storing per-channel data.
+
+Nodes ask their peers to store data using the `update_channel_backup` message:
+
+1. type: 120 (`update_channel_backup`)
+2. data:
+   * [`u16`: `length`]
+   * [`length*byte`:`channel_backup`]
+
+A node with `want_peer_backup_storage` activated:
+  - if its peer doesn't support `provide_peer_backup_storage`:
+    - MUST NOT send `update_channel_backup`
+  - otherwise:
+    - when it sends a message that completes an update of the channel state (`funding_signed`, `commitment_signed`, `revoke_and_ack`, `shutdown` or `closing_signed`):
+      - MAY send an `update_channel_backup` *before* the channel update message
+      - MUST limit its `channel_backup` to 32000 bytes
+    - when it receives `channel_reestablish` with an outdated or missing `channel_backup`:
+      - MAY send a warning
+
+A node with `provide_peer_backup_storage` activated:
+  - when it receives `update_channel_backup`:
+    - if the `channel_backup` exceeds 32000 bytes:
+      - SHOULD send a warning
+      - MUST NOT store this backup data
+    - otherwise:
+      - SHOULD store this backup data
```

Comment:
why is this not a "MUST"?

Also, in BOLT-01 you write
> However, nodes should not expect
their peers to always have their latest backup available (note that there is no
acknowledgement to ensure a backup has been correctly stored).

and below you write
> The `update_channel_backup` should be sent before the message that updates the
channel state (e.g. `commit_sig`) to ensure that the backup is stored with that
latest update, which can be retransmitted on reconnection. If the backup is
sent after the channel update, there is a risk that the channel update is
committed by the remote peer but not the corresponding backup, which may lead
to data loss.

It might be worth specifying that the backup data should be persisted to disk in a blocking manner, before processing subsequent messages.
In particular, imagine a weird architecture where you have a thread that is responsible for persisting backup data blobs to disk and you keep queueing items onto this thread. As such a thread could lag behind the channel state machine (revack, commit_sig) by multiple messages, and we would lose all guarantees about how recent the latest stored backup is compared to the channel state -- ideally (it should be +/-1)

Reply:
> why is this not a "MUST"?

This was made a `SHOULD` based on one of rusty's earlier comment that since this feature is free, you cannot expect your peer to provide a perfect SLA, it's a best effort on their part. In practice I agree with you that for many cases you will really expect your peer to perfectly store your backups (because you're indirectly paying them by connecting to them and creating payment traffic through them), so I don't know whether we should make it a `MUST` or not.

> It might be worth specifying that the backup data should be persisted to disk in a blocking manner, before processing subsequent messages.

Good idea, I'll add something like that.

Reply:
I have slightly extended the requirement in https://github.com/lightning/bolts/pull/881/commits/cac404d2eb7c43b6615c5a0cf798738a57868421, if you have a better wording suggestion don't hesitate to share it!

Reply:
> This was made a SHOULD based on one of rusty's earlier comment that since this feature is free, you cannot expect your peer to provide a perfect SLA, it's a best effort on their part. 

Ah I see, rusty said:
> Require only best effort, not require them to perform a db commit on every new update.

So is it a DOS concern that a channel-peer can send many `channel_backup` messages, forcing us to do lots of db commits? If so, we could maybe relax the requirement to say that next db commit that includes channel-state-changes, must also include the latest received  `channel_backup` blob. i.e. the blob does not have to be persisted as soon as received, but it must be persisted together with the next channel-state-change.

Reply:
> the blob does not have to be persisted as soon as received, but it must be persisted together (at the latest) with the next channel-state-change.

Do you have a specific sentence in mind to to express that requirement? Otherwise I'll try something.

Reply:
I like `channel state change`, let's see what other people think. I included this and rebased in https://github.com/lightning/bolts/pull/881/commits/daa853ce357203ddebb0bc988ac54b9af60fb140